tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h

```diff
@@ -0,0 +1,155 @@
+                  size_t body_buffer_width, int axis_direction, BidirectionalBuffer& buffer) : 
+            BaseDynamics<void>(real_body), LocalDynamics(real_body),GeneralDataDelegateSimple(real_body),
+            buffer_particle_list_(buffer_particle_list), aligned_box_(*shape_ptr.get()),
+            fluid_(DynamicCast<Fluid>(this, particles_->getBaseMaterial())), 
+            pos_n_(particles_->pos_), rho_n_(particles_->rho_), p_(*particles_->getVariableByName<Real>("Pressure")), 
+            axis_(axis_direction), body_buffer_width_(body_buffer_width),                                                                                                          
+            previous_surface_indicator_(*particles_->getVariableByName<int>("PreviousSurfaceIndicator")),                                                                                                                 
+            buffer_particle_indicator_(*particles_->getVariableByName<int>("BufferParticleIndicator")),
+            buffer_(buffer)           
+        {
+            size_t total_body_buffer_particles = 1000.0 * body_buffer_width_;
+            particles_->addBufferParticles(total_body_buffer_particles);
+            real_body.allocateConfigurationMemoriesForBufferParticles();
+        };
+        virtual ~Injection(){};
+
+        /** This class is only implemented in sequential due to memory conflicts. */
+        virtual void update(size_t index_i, Real dt = 0.0);
+
+        virtual void exec(Real dt = 0.0) override
+        {
+            setupDynamics(dt);
+
+            particle_for(ExecutionPolicy(), buffer_particle_list_.size(),
+                         [&](size_t index_i)
+                         { update(index_i, dt); });
+
+            buffer_particle_list_.clear();
+        };
+
+      protected:
+        std::mutex mutex_switch_to_real_;
+        ConcurrentIndexVector &buffer_particle_list_;
+        AlignedBoxShape &aligned_box_;
+        Fluid &fluid_;
+        StdLargeVec<Vecd> &pos_n_;
+        StdLargeVec<Real> &rho_n_, &p_;
+        const int axis_; 
+        size_t body_buffer_width_;
+        StdLargeVec<int> &previous_surface_indicator_, &buffer_particle_indicator_;
+
+      private:
+        BidirectionalBuffer &buffer_;
+    };
+
+  public:
+    BidirectionalBuffer(RealBody &real_body, SharedPtr<AlignedBoxShape> shape_ptr,
+                        size_t body_buffer_width, int axis_direction) : 
+        tag_buffer_particles(this->buffer_particle_list, real_body, shape_ptr),                                                                                                                     
+        injection(this->buffer_particle_list, real_body, shape_ptr, body_buffer_width, axis_direction, *this)
+    {
+        buffer_particle_list.clear();
+    };
+    virtual ~BidirectionalBuffer(){};
+    
+    TagBufferParticles<execution::ParallelPolicy> tag_buffer_particles;
+    Injection<execution::ParallelPolicy> injection;   
```

Comment:
these two variables should be also ended with under line.

Reply:
this is the subclass name, i will consider about it